safety_shield/src/robot_reach.cc

```diff
@@ -2,100 +2,214 @@
+      double r_2 = reach_lib::Point::norm(before.p2_ - after.p2_) / 2 + alpha_i[i + 1] * s_diff * s_diff / 8 +
+                   robot_capsules_[i].r_;
+      // Final radius is maximum of r_1 and r_2 plus the radius expansion for modelling errors.
+      double radius = std::max(r_1, r_2) + secure_radius_;
+      // Enclosure capsule radius is max of ball around p1 and ball around p2
+      reach_capsules.push_back(reach_lib::Capsule(p_1_k, p_2_k, radius));
     }
     return reach_capsules;
-  } catch (const std::exception &exc) {
+  } catch (const std::exception& exc) {
     spdlog::error("Exception in RobotReach::reach: {}", exc.what());
     return {};
   }
 }
 
-} // namespace safety_shield
\ No newline at end of file
+double RobotReach::maxVelocityOfMotion(const Motion& motion) {
+  calculateAllTransformationMatricesAndCapsules(motion.getAngleRef());
+  double max_capsule_vel = 0;
+  for (int i = 0; i < nb_joints_; ++i) {
+    double capsule_vel = getMaxVelocityOfCapsule(i, motion.getVelocityRef());
+    max_capsule_vel = std::max(max_capsule_vel, capsule_vel);
+  }
+  return max_capsule_vel;
+}
+
+void RobotReach::calculateAllTransformationMatricesAndCapsules(const std::vector<double>& q) {
+  robot_capsules_for_velocity_.clear();
+  z_vectors_.clear();
+  Eigen::Matrix4d T_capsule = transformation_matrices_[0];
+  // push z-vector of transformation matrix
+  z_vectors_.emplace_back(T_capsule.block(0, 2, 3, 1));
+  for (int i = 0; i < nb_joints_; ++i) {
+    forwardKinematic(q[i], i, T_capsule);
+    reach_lib::Capsule capsule = transformCapsule(i, T_capsule);
+    // push robot capsule and z-vector
+    robot_capsules_for_velocity_.push_back(capsule);
+    z_vectors_.emplace_back(T_capsule.block(0, 2, 3, 1));
+  }
+}
+
+RobotReach::CapsuleVelocity RobotReach::getVelocityOfCapsule(const int capsule, std::vector<double> q_dot){
+  Eigen::VectorXd velocity = Eigen::Map<Eigen::VectorXd, Eigen::Unaligned>(q_dot.data(), q_dot.size());
+  // Point 1
+  Eigen::Matrix<double, 6, Eigen::Dynamic> jacobian1 = getJacobian(capsule, robot_capsules_for_velocity_[capsule].p1_);
+  Eigen::Vector<double, 6> result1 = jacobian1 * velocity.segment(0, capsule + 1);
+  RobotReach::SE3Vel vel1(result1.segment(0, 3), result1.segment(3, 3));
+  // Point 2: v_2 = v_1 + \omega_1 \times (p_2 - p_1), \omega_2 = \omega_1
+  RobotReach::SE3Vel vel2(
+    vel1.first + vel1.second.cross(pointTo3dVector(
+      robot_capsules_for_velocity_[capsule].p2_ - robot_capsules_for_velocity_[capsule].p1_
+    )),
+    vel1.second);
+  return RobotReach::CapsuleVelocity(vel1, vel2);
+}
+
+double RobotReach::getMaxVelocityOfCapsule(const int capsule, std::vector<double> q_dot) {
+  RobotReach::CapsuleVelocity capsule_velocity = getVelocityOfCapsule(capsule, q_dot);
+  return std::max(
+    getMaxCartVelocityOfCapsulePoint(capsule, capsule_velocity.first),
+    getMaxCartVelocityOfCapsulePoint(capsule, capsule_velocity.second)
```

Comment:
only calculate velocity for second capsule point, because `approximateVelOfCapsule()` and `exactVelOfCapsule()` take the second capsule point as the parameter. Calculating the velocity for first capsule point is not needed and would save some computations